File: parity-tools/harness/coercion.py
"""Tag-aware JSON coercion for parity harness fixtures.

Converts JSON values tagged with {"__type": "...", "value": ...} into native
Python structures. Plain primitives, lists, and untagged dicts pass through
unchanged (with recursion into their members).
"""
import sys
from pathlib import Path

# Make jhora package importable regardless of invocation cwd.
_REPO_ROOT = Path(__file__).resolve().parents[2]
_SRC = _REPO_ROOT / "src"
if str(_SRC) not in sys.path:
    sys.path.insert(0, str(_SRC))

from jhora.panchanga.drik import Date, Place  # noqa: E402


def _coerce_place(value):
    return Place(value["name"], value["latitude"], value["longitude"], value["timezone"])


def _coerce_date(value):
    return Date(value["year"], value["month"], value["day"])


_COERCERS = {
    "Place": _coerce_place,
    "Date": _coerce_date,
}


def coerce(node):
    """Recursively convert tagged JSON to native Python values."""
    if isinstance(node, dict):
        type_tag = node.get("__type")
        if type_tag is not None:
            coercer = _COERCERS.get(type_tag)
            if coercer is None:
                raise ValueError(f"Unknown __type tag: {type_tag!r}")
            return coercer(node["value"])
        return {k: coerce(v) for k, v in node.items()}
    if isinstance(node, list):
        return [coerce(x) for x in node]
    return node

File: pyjhora-web/src/parity/coercion.ts
/**
 * Tag-aware JSON coercion for parity harness fixtures.
 * Mirror of parity-tools/harness/coercion.py.
 *
 * Converts JSON values tagged with {__type, value} into native TS structures.
 * Plain primitives, arrays, and untagged objects pass through (with recursion).
 */

type Coercer = (value: any) => any;

const coercers: Record<string, Coercer> = {
  Place: (value) => ({
    name: value.name,
    latitude: value.latitude,
    longitude: value.longitude,
    timezone: value.timezone,
  }),
  Date: (value) => ({
    year: value.year,
    month: value.month,
    day: value.day,
  }),
};

export function coerce(node: unknown): unknown {
  if (Array.isArray(node)) {
    return node.map((x) => coerce(x));
  }
  if (node !== null && typeof node === 'object') {
    const obj = node as Record<string, unknown>;
    const typeTag = obj.__type as string | undefined;
    if (typeTag !== undefined) {
      const coercer = coercers[typeTag];
      if (coercer === undefined) {
        throw new Error(`Unknown __type tag: ${typeTag}`);
      }
      return coercer(obj.value);
    }
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(obj)) {
      out[k] = coerce(v);
    }
    return out;
  }
  return node;
}
